akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2.scala

```diff
@@ -84,13 +84,15 @@ private[http] final class Http2Ext(implicit val system: ActorSystem)
         incoming: Tcp.IncomingConnection =>
           try {
             httpPlusSwitching(http1, http2).addAttributes(prepareServerAttributes(settings, incoming))
-              .watchTermination()(Keep.both)
-              .join(HttpConnectionIdleTimeoutBidi(settings.idleTimeout, Some(incoming.remoteAddress)) join incoming.flow)
-              .mapMaterializedValue {
-                case (connectionTerminator, future) =>
-                  connectionTerminator.foreach(masterTerminator.registerConnection(_)(fm.executionContext))((fm.executionContext))
+              .watchTermination() {
+                case (connectionTerminatorF, future) =>
+                  connectionTerminatorF.foreach { connectionTerminator =>
+                    masterTerminator.registerConnection(connectionTerminator)(fm.executionContext)
+                    future.onComplete(_ => masterTerminator.removeConnection(connectionTerminator))(fm.executionContext)
```

Comment:
`onComplete` :+1: 

Perhaps we should add a comment to `registerConnection`, but would remain easy to miss...

Reply:
Or we could use WeakReferences instead, though, not sure if that's really necessary.

Reply:
Or a protocol between MasterServerTerminator and ServerTerminator, so that a ServerTerminator knows about the MasterServerTerminator and could unregister itself on `postStop`.